### ISSUE TITLE
Okta API Issues

### ISSUE THREAD

**Issue by quantumew** [articulate/terraform-provider-okta#316]
Suddenly our tests are failing and it is for a reason out of our control. I can watch an App get created, return a 200 on the first read, then start 404ing on all subsequent calls, then when I go to debug suddenly the app starts 200ing again. I would guess this is a cache issue on Okta end.

https://support.okta.com/help/s/case/5001Y00001K9bnEQAR

**Comment by quantumew** [comment]
Debug logs showing POST works, GET on app and app groups is OK but app user listing endpoint 404. I will get a 404 on this endpoint for over a second after getting a 200 on other endpoints. Seems to definitely be a caching bug on Okta's API.

```
POST /api/v1/apps?activate=true                                                Xb9s8UMxMNGkJAz24Le6vQAABpQ                   18:12:33             200
GET  /api/v1/apps/0oao8ybrjiONWB8U20h7/groups                   Xb9s8ihPVadaOZrolZd5SgAACj4                          18:12:34             200
GET  /api/v1/apps/0oao8ybrjiONWB8U20h7/users                     Xb9s83PIOQtNJ5grq@GGxQAAAEc                    18:12:34              404
GET  /api/v1/apps/0oao8ybrjiONWB8U20h7                               Xb9s86yLHRZ16MUHVXgWXAAACMA                18:12:35              200
GET  /api/v1/apps/0oao8ybrjiONWB8U20h7/users                      Xb9s8xC0pFCa7qcuU4T4LwAADKE                     18:12:35             404
```

**Comment by jpf** [comment]
Hey @quantumew - I'm seeing this issue too and am actively looking into it.

**Comment by quantumew** [comment]
@jpf awesome, really appreciate that! Let me know what you find and if it will require any changes on my end to work around it.

**Comment by jeremie0** [comment]
I'm seeing this too.  sometimes it works to delete the newly created app from the UI and then re-create via terraform but that is not consistent.

**Comment by jpf** [comment]
Hey @quantumew - I'm able to see errors in the logs, but I'm having trouble replicating the error "by hand" - here's what I'm running:

```
#!/usr/bin/env bash

id=$(curl -X POST "https://${OKTA_URL}/api/v1/apps" \
  -H 'Accept: application/json' \
  -H "Authorization: SSWS ${OKTA_API_TOKEN}" \
  -H 'Content-Type: application/json' \
  -d '{
  "name": "oidc_client",
  "label": "Test app",
  "status": "ACTIVE",
  "signOnMode": "OPENID_CONNECT",
  "settings": {
      "oauthClient": {
          "client_uri": "http://localhost:8080",
          "redirect_uris": [
              "https://example.com/oauth2/callback",
              "myapp://callback"
          ],
          "response_types": [
              "token",
              "id_token",
              "code"
          ],
          "grant_types": [
              "implicit",
              "authorization_code"
          ],
          "application_type": "native"
      }
  }
}' | jq -r '.id')

echo "Got id: ${id}"

sleep 3

curl -v -X GET "https://${OKTA_URL}/api/v1/apps/${id}" \
  -H 'Accept: application/json' \
  -H "Authorization: SSWS ${OKTA_API_TOKEN}" \
  -H 'Content-Type: application/json' | jq

curl -v -X GET "https://${OKTA_URL}/api/v1/apps/${id}/users" \
  -H 'Accept: application/json' \
  -H "Authorization: SSWS ${OKTA_API_TOKEN}" \
  -H 'Content-Type: application/json' | jq
```

**Comment by quantumew** [comment]
That is exactly what I used to repro this issue yesterday, it was a toss up whether it would work or not even with a 5 second sleep. Today I am seeing it less. If you remove the sleep and run it a few times it will happen for you. I added the sleep in yesterday to demonstrate that even with a long sleep it would still happen.

**Comment by quantumew** [comment]
Received an update on my case, `Since we have reported this issue to Engineering,  the initial investigation revealed a replication issue on our database that caused a delay for the preview cell (OP1) with data replication across the database.`. Some initial testing is showing an improvement. Although I did see this error once in the wild after receiving this update so I will run the acceptance test throughout the day, and report back.
